src/server/journal/journal_slice.cc

```diff
@@ -116,36 +117,63 @@ error_code JournalSlice::Close() {
 }
 #endif
 
+bool JournalSlice::IsLSNInBuffer(LSN lsn) const {
+  DCHECK(ring_buffer_);
+
+  if (ring_buffer_->empty()) {
+    return false;
+  }
+  return (*ring_buffer_)[0].lsn <= lsn && lsn <= ((*ring_buffer_)[ring_buffer_->size() - 1].lsn);
+}
+
+std::string_view JournalSlice::GetEntry(LSN lsn) const {
+  DCHECK(ring_buffer_ && IsLSNInBuffer(lsn));
+  auto start = (*ring_buffer_)[0].lsn;
+  DCHECK((*ring_buffer_)[lsn - start].lsn == lsn);
+  return (*ring_buffer_)[lsn - start].data;
+}
+
 void JournalSlice::AddLogRecord(const Entry& entry, bool await) {
-  // DCHECK(ring_buffer_);
-  if (entry.opcode != Op::NOOP) {
-    lsn_++;
-// TODO: This is preparation for AOC style journaling, currently unused.
-#if 0
-    RingItem item;
-    item.lsn = prev_lsn;
+  DCHECK(ring_buffer_);
+
+  JournalItem dummy;
+  JournalItem* item;
+  if (entry.opcode == Op::NOOP) {
+    item = &dummy;
+    item->lsn = -1;
+    item->opcode = entry.opcode;
+    item->data = "";
+  } else {
+    item = ring_buffer_->GetTail(true);
+    item->opcode = entry.opcode;
+    item->lsn = lsn_++;
 
-    item.opcode = entry.opcode;
-    item.txid = entry.txid;
-    VLOG(1) << "Writing item [" << item.lsn << "]: " << entry.ToString();
-    ring_buffer_->EmplaceOrOverride(move(item));
+    io::BufSink buf_sink{&ring_serialize_buf_};
+    JournalWriter writer{&buf_sink};
+    writer.Write(entry);
 
+    item->data = io::View(ring_serialize_buf_.InputBuffer());
+    ring_serialize_buf_.Clear();
+    VLOG(1) << "Writing item [" << item->lsn << "]: " << entry.ToString();
```

Comment:
as a rule of thumb, `VLOG(1)` is for infrequent operations in a file, `VLOG(2)` is for logs `per item` or `per event`

Reply:
changed to 2